### DISCUSSION TITLE
Component Level Selective Rendering || Component Level Cache Based on Data Life

### DISCUSSION THREAD

**Discussion by MuhammadKhizar7** [nuxt/nuxt#16469] [Ideas]
Route base cache is great, but I am wondering if it could be possible to use component level cache like in image I tried to make it easy to understand, different components have different life span based on their data like header and footer are mostly static and 4th and 5th sections can be cached for a day or week, only thing that changes more often is card (I tried to make use case like ecommerce or real estate etc.)
I might be wrong but if only component like cards section changes whole page need to rerender and need fetch data of whole page like 

- one request for navbar 
- one for hero section
- one for cards 
- one for locations/categories
- one for footer data

One solution is to cache in Redis or some other caching solution, but it's still needed to render whole page 
and still needed to fetch data in case of edge function or Cloudflare workers there is time limitation or also cost (CPU/sec) to render (fetch data and rendering time) -> just for example 

What I think (I can be wrong) is can be a solution is component level selective rendering rules like for example header or layout component mostly not changes so can be static render for x amount of time and other component patch it up based on their rules
For example in image only component that changes more frequently is cards section so only this section render and patch up with prerander page on server or in case of client for example dashboard only client component page up but not whole dashboard

![Nuxt_component_cache_idea-2022-11-05-2139](https://user-images.githubusercontent.com/44477395/200134008-30d3dae2-fd0f-4a61-8bc4-847584c7ac2f.png)

**Comment by milelazar** [comment]
It has been a long time but i am wondering if you figured this out? I am currently searching for the same thing. I have thousands of products and route cache is great but the same data is fetched on every new product route so it affects server performance.

There is Nuxt Multi Cache module that supports component caching. But if you have nested components or something more complex, it does not work very well.

**Reply by s0h311** [reply]
@milelazar for me it sounds like you fire unnecessary fetches, which can be avoided. Correct me if I'm wrong but vue.js uses the concept of signals, which eliminates wasteful rerenders. It means only components whose (reactive-) data is changed get rerendered. Therefore perhaps the question you want to ask is, how do you fetch the data? Do you cache it after fetching?

**Reply by milelazar** [reply]
I have redis cached all data endpoints and nitro routes so this should help. There is around 1000-1200 users active at all times (not sure how many resources bots use in addition). There are tens of thousands of products in addition to blog posts and other cms pages. So it is very likely that new user will hit a page/product for the first time that is not yet cached.

Once you open the site for the first time and click around, there are no additional duplicate fetches. So component rerendering and refetching should not be a problem.

Server cpu load is around 65-70% on vps hosting. It looks high so im thinking that component caching might be helpful. If menu component with single fetch is cached and served for every new user, there should be a lot less fetches. With route caching, that menu will be fetched for every new user. So 500 times if there are 500 new users on a page that is not yet route cached. 
Or maybe i am completely wrong. Please let me know if i am.
